### PR TITLE
_lpdf-compatible reduce_sum

### PR DESCRIPTION
**Pull request by nhuurre**
companion to https://github.com/stan-dev/math/pull/1843

Userdef `_lpdf` are now allowed in `reduce_sum`. Also fixes `reduce_sum_generic` error message that incorrectly claimed that grainsize argument is before the array.

### CODE REVIEW

File: test/integration/bad/reduce_sum/bad_return_type_static.abc [hunk 1]
```diff
@@ -1,5 +1,5 @@
 functions {
-  int my_func_0(int start, int end, real[] y_slice) {
+  int my_func_0(real[] y_slice, int start, int end) {
```
Comment (rok-cesnovar) [inline test/integration/bad/reduce_sum/bad_return_type_static.abc:2]: This one is on me, but can you change the extension. Dont know why I put .abc :)

**Comment by nhuurre** [comment]
There's an `_lpdf` user function in test/integration/good/code-gen/reduce_sum_m1.stan. Is that not enough?

**Comment by rok-cesnovar** [comment]
Ah missed that sorry. Yes, that is plenty enough. We can also fix the file later, not a huge deal.

**Comment by rok-cesnovar** [comment]
I will merge this in a few hours, once the math submodule change is propagated to the cmdstan level on develop (merge tests of the math PR need to pass).

**Pull request merged**